File: backend/app/tasks/manager.py
"""任务管理器：线程 worker + 线程安全队列 + SQLite 持久化。

FastAPI 同步接口在线程池执行，任务入队来自任意线程；
使用 queue.Queue（线程安全）而非 asyncio.Queue，避免跨线程唤醒问题。
任务级错误隔离：单个任务失败只影响自己，不中断队列与其他平台任务。
"""
from __future__ import annotations  # 类体内方法注解延迟求值，避免 list 等被类内方法遮蔽

import queue
import re
import threading
import time
import uuid
from pathlib import Path

import httpx

from ..config import load_config
from .db import DB

UA = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36"
)


def detect_platform(url: str) -> str:
    u = url.lower()
    if "douyin.com" in u or "iesdouyin" in u or "v.douyin" in u:
        return "douyin"
    if "x.com" in u or "twitter.com" in u:
        return "x"
    return ""


def _sanitize_filename(name: str) -> str:
    name = re.sub(r'[\\/:*?"<>|\r\n]', "_", name).strip() or "untitled"
    return name[:60]


def _referer_for(url: str) -> str:
    """按 CDN 域名选择合法 Referer，避免陌生来源被拒（抖音/ X 各自主机校验）。"""
    if "twimg.com" in url:
        return "https://x.com/"
    return "https://www.douyin.com/"


class TaskManager:
    def __init__(self):
        self.db = DB()
        self.queue: queue.Queue = queue.Queue()
        self.worker: threading.Thread | None = None
        self._tasks: dict[str, dict] = {}
        self._lock = threading.Lock()
        self._load_from_db()

    # ---- 生命周期 ----

    def start(self) -> None:
        if self.worker is None or not self.worker.is_alive():
            self.worker = threading.Thread(target=self._run, name="task-worker", daemon=True)
            self.worker.start()

    def stop(self) -> None:
        pass  # 守护线程随进程退出

    # ---- 外部接口 ----

    def create(self, task_type: str = "link", url: str | None = None, options: dict | None = None) -> dict:
        tid = uuid.uuid4().hex[:12]
        created_at = time.strftime("%Y-%m-%d %H:%M:%S")
        info = {
            "id": tid, "type": task_type, "status": "pending", "progress": 0,
            "output": None, "error": None, "created_at": created_at,
            "url": url, "options": options or {}, "cancelled": False,
        }
        with self._lock:
            self._tasks[tid] = info
        self.db.upsert(tid, type=task_type, status="pending", progress=0,
                       created_at=created_at, options=options or {})
        self.queue.put(tid)
        return self.public_view(tid)

    def get(self, tid: str) -> dict | None:
        return self.public_view(tid)

    def list(self) -> list[dict]:
        with self._lock:
            ids = sorted(self._tasks, key=lambda t: self._tasks[t]["created_at"], reverse=True)
        return [self.public_view(tid) for tid in ids]

    def cancel(self, tid: str) -> bool:
        with self._lock:
            info = self._tasks.get(tid)
            if not info or info["status"] not in ("pending", "running"):
                return False
            info["cancelled"] = True
        return True

    def public_view(self, tid: str) -> dict | None:
        with self._lock:
            info = self._tasks.get(tid)
            if not info:
                return None
            return {
                "id": info["id"], "type": info["type"], "status": info["status"],
                "progress": info["progress"], "output": info["output"],
                "error": info["error"], "created_at": info["created_at"],
            }

    def _update(self, tid: str, **fields) -> None:
        with self._lock:
            info = self._tasks.get(tid)
            if info:
                for k, v in fields.items():
                    if k in ("status", "progress", "output", "error"):
                        info[k] = v
        self.db.update(tid, **fields)

    # ---- 内部 ----

    def _load_from_db(self) -> None:
        for row in self.db.list_all():
            if row["status"] in ("pending", "running"):
                row["status"] = "failed"
                row["error"] = "应用重启，任务中断"
                self.db.update(row["id"], status="failed", error=row["error"])
            with self._lock:
                self._tasks[row["id"]] = {**row, "url": None, "options": {}, "cancelled": False}

    def _run(self) -> None:
        while True:
            tid = self.queue.get()
            try:
                with self._lock:
                    info = self._tasks.get(tid)
                if info and not info["cancelled"]:
                    self._process(tid)
            finally:
                self.queue.task_done()

    def _process(self, tid: str) -> None:
        self._update(tid, status="running", progress=5)
        try:
            with self._lock:
                info = self._tasks[tid]
                url = info.get("url") or ""
                options = info.get("options") or {}
            cfg = load_config()
            out_dir = Path(options.get("output_dir") or cfg["output_dir"])
            out_dir.mkdir(parents=True, exist_ok=True)

            if info["type"] == "direct":
                # 直链下载：url 为素材直链，options 携带 kind/title/image_url
                kind = options.get("kind") or "video"
                title = options.get("title") or "untitled"
                out = self._download(
                    url, out_dir, title, kind, tid,
                    image_url=options.get("image_url"),
                )
                output = " | ".join(map(str, out)) if isinstance(out, list) else str(out)
                self._update(tid, status="done", progress=100, output=output)
                return

            # link 任务：解析后下载全部素材
            platform = detect_platform(url)
            if not platform:
                raise RuntimeError("无法识别的平台链接")
            from ..parsers import PARSERS
            parser = PARSERS.get(platform)
            if parser is None:
                raise RuntimeError(f"「{platform}」平台解析尚未支持（X 平台计划在 M4 里程碑接入）")

            result = parser.parse(url, options.get("remove_platform_wm", True))
            self._update(tid, progress=15)

            if not result.files:
                raise RuntimeError("解析结果为空")
            total = len(result.files)
            paths = []
            for i, f in enumerate(result.files):
                start_pct = 15 + int(i / total * 80)
                end_pct = 15 + int((i + 1) / total * 80)
                name = result.title if total == 1 else f"{result.title}_{i + 1}"
                out = self._download(
                    f.url, out_dir, name, f.kind, tid, start_pct, end_pct,
                    image_url=f.image_url,
                )
                if isinstance(out, list):
                    paths.extend(map(str, out))
                else:
                    paths.append(str(out))
            self._update(tid, status="done", progress=100, output=" | ".join(paths))
        except Exception as exc:
            with self._lock:
                cancelled = self._tasks.get(tid, {}).get("cancelled", False)
            if cancelled:
                self._update(tid, status="cancelled", error="用户取消")
            else:
                self._update(tid, status="failed", error=str(exc))

    def _download(
        self, url: str, out_dir: Path, title: str, kind: str, tid: str,
        start_pct: int = 15, end_pct: int = 99, image_url: str | None = None,
    ) -> Path | list[Path]:
        """下载单个素材；Live 图（带 image_url）下载「静态照片 + 视频」双文件。"""
        if kind == "livephoto" and image_url:
            mid = start_pct + (end_pct - start_pct) * 2 // 3
            video_path = self._download(url, out_dir, title, "video", tid, start_pct, mid)
            photo_path = self._download(image_url, out_dir, title, "image", tid, mid, end_pct)
            return [video_path, photo_path]
        ext = ".mp4" if kind in ("video", "livephoto") else ".gif" if kind == "gif" else self._ext_from_url(url)
        out_path = out_dir / f"{_sanitize_filename(title)}_{time.strftime('%Y%m%d%H%M%S')}{ext}"
        total = 0
        with httpx.Client(
            headers={"User-Agent": UA, "Referer": _referer_for(url)},
            follow_redirects=True,
            timeout=120,
        ) as client:
            with client.stream("GET", url) as resp:
                resp.raise_for_status()
                content_length = int(resp.headers.get("content-length") or 0)
                with open(out_path, "wb") as f:
                    for chunk in resp.iter_bytes():
                        with self._lock:
                            cancelled = self._tasks[tid].get("cancelled", False)
                        if cancelled:
                            raise RuntimeError("cancelled")
                        f.write(chunk)
                        total += len(chunk)
                        if content_length:
                            pct = start_pct + int(total / content_length * (end_pct - start_pct))
                            self._update(tid, progress=max(start_pct, min(end_pct, pct)))
        return out_path

    @staticmethod
    def _ext_from_url(url: str) -> str:
        """按直链路径推断扩展名，避免 .webp 等格式存成 .jpg。"""
        path = url.split("?")[0].lower()
        for e in (".webp", ".png", ".gif", ".jpeg", ".jpg"):
            if path.endswith(e):
                return e
        return ".jpg"


manager = TaskManager()

File: backend-rust/src/tasks/manager.rs
//! 任务管理器：tokio worker + 异步队列 + SQLite 持久化。
//! 逻辑与 Python 版 backend/app/tasks/manager.py 一致（线程 worker → tokio 异步等价）。

use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};
use std::time::Duration;

use chrono::Local;
use futures_util::StreamExt;
use reqwest::header::{HeaderMap, HeaderValue, REFERER, USER_AGENT};
use serde_json::Value;
use tokio::io::AsyncWriteExt;
use tokio::sync::mpsc;
use uuid::Uuid;

use super::db::DB;
use crate::config::load_config;
use crate::schemas::TaskRead;

pub const UA: &str = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36";

/// 识别链接所属平台（与 Python detect_platform 一致）。
pub fn detect_platform(url: &str) -> &'static str {
    let u = url.to_lowercase();
    if u.contains("douyin.com") || u.contains("iesdouyin") || u.contains("v.douyin") {
        "douyin"
    } else if u.contains("x.com") || u.contains("twitter.com") {
        "x"
    } else {
        ""
    }
}

/// 按 CDN 域名选择合法 Referer（与 Python _referer_for 一致）。
pub fn _referer_for(url: &str) -> &'static str {
    if url.contains("twimg.com") {
        return "https://x.com/";
    }
    "https://www.douyin.com/"
}

fn _sanitize_filename(name: &str) -> String {
    let cleaned: String = name
        .chars()
        .map(|c| match c {
            '\\' | '/' | ':' | '*' | '?' | '"' | '<' | '>' | '|' | '\r' | '\n' => '_',
            c => c,
        })
        .collect();
    let t = cleaned.trim();
    if t.is_empty() {
        "untitled".to_string()
    } else {
        t.chars().take(60).collect()
    }
}

fn _ext_from_url(url: &str) -> &'static str {
    let path = url.split('?').next().unwrap_or("").to_lowercase();
    for e in [".webp", ".png", ".gif", ".jpeg", ".jpg"] {
        if path.ends_with(e) {
            return e;
        }
    }
    ".jpg"
}

#[derive(Clone)]
struct TaskInner {
    id: String,
    task_type: String,
    status: String,
    progress: i32,
    output: Option<String>,
    error: Option<String>,
    created_at: String,
    url: Option<String>,
    options: Value,
    cancelled: bool,
}

pub struct TaskManager {
    db: DB,
    tx: mpsc::Sender<String>,
    tasks: Mutex<HashMap<String, TaskInner>>,
}

impl TaskManager {
    pub fn new() -> Arc<Self> {
        let (tx, rx) = mpsc::channel::<String>(1024);
        let mgr = Arc::new(TaskManager {
            db: DB::new(),
            tx,
            tasks: Mutex::new(HashMap::new()),
        });
        mgr.load_from_db();
        let worker = mgr.clone();
        tokio::spawn(async move {
            worker.run(rx).await;
        });
        mgr
    }

    // ---- 外部接口（与 Python 一致） ----

    pub async fn create(
        self: &Arc<Self>,
        task_type: &str,
        url: Option<&str>,
        options: Value,
    ) -> TaskRead {
        let tid = Uuid::new_v4().simple().to_string()[..12].to_string();
        let created_at = Local::now().format("%Y-%m-%d %H:%M:%S").to_string();
        let info = TaskInner {
            id: tid.clone(),
            task_type: task_type.to_string(),
            status: "pending".to_string(),
            progress: 0,
            output: None,
            error: None,
            created_at: created_at.clone(),
            url: url.map(|s| s.to_string()),
            options,
            cancelled: false,
        };
        let db_options = info.options.clone();
        {
            let mut tasks = self.tasks.lock().unwrap();
            tasks.insert(tid.clone(), info);
        }
        let _ = self.db.upsert(
            &tid,
            task_type,
            "pending",
            0,
            None,
            None,
            &created_at,
            &db_options,
        );
        let _ = self.tx.send(tid.clone()).await;
        self.public_view(&tid).unwrap()
    }

    pub fn get(&self, tid: &str) -> Option<TaskRead> {
        self.public_view(tid)
    }

    pub fn list(&self) -> Vec<TaskRead> {
        let tasks = self.tasks.lock().unwrap();
        let mut ids: Vec<&TaskInner> = tasks.values().collect();
        ids.sort_by(|a, b| b.created_at.cmp(&a.created_at));
        ids.into_iter().map(|t| t.into()).collect()
    }

    pub fn cancel(&self, tid: &str) -> bool {
        let mut tasks = self.tasks.lock().unwrap();
        match tasks.get_mut(tid) {
            Some(info) if info.status == "pending" || info.status == "running" => {
                info.cancelled = true;
                true
            }
            _ => false,
        }
    }

    pub fn public_view(&self, tid: &str) -> Option<TaskRead> {
        let tasks = self.tasks.lock().unwrap();
        tasks.get(tid).map(|t| t.into())
    }

    fn update(&self, tid: &str, status: Option<&str>, progress: Option<i32>, output: Option<&str>, error: Option<&str>) {
        {
            let mut tasks = self.tasks.lock().unwrap();
            if let Some(info) = tasks.get_mut(tid) {
                if let Some(v) = status {
                    info.status = v.to_string();
                }
                if let Some(v) = progress {
                    info.progress = v;
                }
                if let Some(v) = output {
                    info.output = Some(v.to_string());
                }
                if let Some(v) = error {
                    info.error = Some(v.to_string());
                }
            }
        }
        let _ = self.db.update(tid, status, progress, output, error);
    }

    // ---- 内部 ----

    fn load_from_db(&self) {
        let rows = self.db.list_all().unwrap_or_default();
        let mut tasks = self.tasks.lock().unwrap();
        for mut row in rows {
            if row.status == "pending" || row.status == "running" {
                row.status = "failed".to_string();
                let err = Some("应用重启，任务中断".to_string());
                row.error = err.clone();
                let _ = self.db.update(&row.id, Some("failed"), None, None, err.as_deref());
            }
            tasks.insert(
                row.id.clone(),
                TaskInner {
                    id: row.id,
                    task_type: row.task_type,
                    status: row.status,
                    progress: row.progress,
                    output: row.output,
                    error: row.error,
                    created_at: row.created_at,
                    url: None,
                    options: serde_json::json!({}),
                    cancelled: false,
                },
            );
        }
    }

    async fn run(self: Arc<Self>, mut rx: mpsc::Receiver<String>) {
        while let Some(tid) = rx.recv().await {
            let cancelled = {
                let tasks = self.tasks.lock().unwrap();
                tasks.get(&tid).map(|t| t.cancelled).unwrap_or(false)
            };
            if !cancelled {
                self.process(&tid).await;
            }
        }
    }

    async fn process(&self, tid: &str) {
        self.update(tid, Some("running"), Some(5), None, None);
        let outcome = self.process_inner(tid).await;
        match outcome {
            Ok(()) => {}
            Err((cancelled, err)) => {
                if cancelled {
                    self.update(tid, Some("cancelled"), None, None, Some("用户取消"));
                } else {
                    self.update(tid, Some("failed"), None, None, Some(&err));
                }
            }
        }
    }

    async fn process_inner(&self, tid: &str) -> Result<(), (bool, String)> {
        let (url, options, task_type) = {
            let tasks = self.tasks.lock().unwrap();
            let info = tasks.get(tid).unwrap();
            (info.url.clone(), info.options.clone(), info.task_type.clone())
        };
        let url = url.unwrap_or_default();
        let cfg = load_config();
        let out_dir = options
            .get("output_dir")
            .and_then(|v| v.as_str())
            .filter(|s| !s.is_empty())
            .map(PathBuf::from)
            .unwrap_or_else(|| PathBuf::from(&cfg.output_dir));
        std::fs::create_dir_all(&out_dir).map_err(|e| (false, e.to_string()))?;

        if task_type == "direct" {
            // 直链下载：url 为素材直链，options 携带 kind/title/image_url
            let kind = options.get("kind").and_then(|v| v.as_str()).unwrap_or("video").to_string();
            let title = options.get("title").and_then(|v| v.as_str()).unwrap_or("untitled").to_string();
            let out = self
                .download(&url, &out_dir, &title, &kind, tid, 15, 99, options.get("image_url").and_then(|v| v.as_str()))
                .await?;
            let output = out.iter().map(|p| p.to_string_lossy().to_string()).collect::<Vec<_>>().join(" | ");
            self.update(tid, Some("done"), Some(100), Some(&output), None);
            return Ok(());
        }

        // link 任务：解析后下载全部素材（M1 骨架解析尚未接入，将返回占位错误）
        let platform = detect_platform(&url);
        if platform.is_empty() {
            return Err((false, "无法识别的平台链接".to_string()));
        }
        let remove_platform_wm = options
            .get("remove_platform_wm")
            .and_then(|v| v.as_bool())
            .unwrap_or(true);
        let result = crate::parsers::parse(platform, &url, remove_platform_wm)
            .await
            .map_err(|e| (self.is_cancelled(tid), e))?;
        self.update(tid, None, Some(15), None, None);

        if result.files.is_empty() {
            return Err((self.is_cancelled(tid), "解析结果为空".to_string()));
        }
        let total = result.files.len();
        let mut paths: Vec<String> = Vec::new();
        for (i, f) in result.files.iter().enumerate() {
            let start_pct = 15 + (i * 80) / total;
            let end_pct = 15 + ((i + 1) * 80) / total;
            let name = if total == 1 {
                result.title.clone()
            } else {
                format!("{}_{}", result.title, i + 1)
            };
            let out = self
                .download(&f.url, &out_dir, &name, &f.kind, tid, start_pct as i32, end_pct as i32, f.image_url.as_deref())
                .await?;
            paths.extend(out.iter().map(|p| p.to_string_lossy().to_string()));
        }
        self.update(tid, Some("done"), Some(100), Some(&paths.join(" | ")), None);
        Ok(())
    }

    async fn download(
        &self,
        url: &str,
        out_dir: &Path,
        title: &str,
        kind: &str,
        tid: &str,
        start_pct: i32,
        end_pct: i32,
        image_url: Option<&str>,
    ) -> Result<Vec<PathBuf>, (bool, String)> {
        // Live 图（带 image_url）：下载「静态照片 + 视频」双文件
        if kind == "livephoto" {
            if let Some(img) = image_url {
                let mid = start_pct + (end_pct - start_pct) * 2 / 3;
                let video = Box::pin(self.download(url, out_dir, title, "video", tid, start_pct, mid, None)).await?;
                let photo = Box::pin(self.download(img, out_dir, title, "image", tid, mid, end_pct, None)).await?;
                let mut paths = video;
                paths.extend(photo);
                return Ok(paths);
            }
        }
        let ext = if kind == "video" || kind == "livephoto" {
            ".mp4"
        } else if kind == "gif" {
            ".gif"
        } else {
            _ext_from_url(url)
        };
        let filename = format!(
            "{}_{}{}",
            _sanitize_filename(title),
            Local::now().format("%Y%m%d%H%M%S"),
            ext
        );
        let out_path = out_dir.join(filename);

        let mut headers = HeaderMap::new();
        headers.insert(USER_AGENT, HeaderValue::from_static(UA));
        headers.insert(REFERER, HeaderValue::from_static(_referer_for(url)));

        let client = reqwest::Client::builder()
            .redirect(reqwest::redirect::Policy::default())
            .timeout(Duration::from_secs(120))
            .build()
            .map_err(|e| (false, e.to_string()))?;
        let resp = client
            .get(url)
            .headers(headers)
            .send()
            .await
            .map_err(|e| (false, format!("下载请求失败：{e}")))?;
        let status = resp.status();
        if !status.is_success() {
            return Err((false, format!("下载失败：HTTP {}", status)));
        }
        let total_len = resp.content_length().unwrap_or(0);
        let mut file = tokio::fs::File::create(&out_path)
            .await
            .map_err(|e| (false, format!("创建文件失败：{e}")))?;
        let mut stream = resp.bytes_stream();
        let mut written: u64 = 0;
        while let Some(chunk) = stream.next().await {
            let chunk = chunk.map_err(|e| (false, format!("读取数据失败：{e}")))?;
            let cancelled = self.is_cancelled(tid);
            if cancelled {
                return Err((true, "cancelled".to_string()));
            }
            file.write_all(&chunk)
                .await
                .map_err(|e| (false, format!("写入文件失败：{e}")))?;
            written += chunk.len() as u64;
            if total_len > 0 {
                let pct = start_pct + (written as f64 / total_len as f64 * (end_pct - start_pct) as f64) as i32;
                let pct = pct.clamp(start_pct, end_pct);
                self.update(tid, None, Some(pct), None, None);
            }
        }
        Ok(vec![out_path])
    }

    fn is_cancelled(&self, tid: &str) -> bool {
        self.tasks
            .lock()
            .unwrap()
            .get(tid)
            .map(|t| t.cancelled)
            .unwrap_or(false)
    }
}

impl From<&TaskInner> for TaskRead {
    fn from(t: &TaskInner) -> Self {
        TaskRead {
            id: t.id.clone(),
            task_type: t.task_type.clone(),
            status: t.status.clone(),
            progress: t.progress,
            output: t.output.clone(),
            error: t.error.clone(),
            created_at: t.created_at.clone(),
        }
    }
}
